### PR TITLE
Update continent query with input from latest team quality review

### PR DESCRIPTION
**Pull request by critzo**
This PR updates [one statistics query](https://github.com/m-lab/stats-pipeline/blob/random-sampling/statistics/queries/continent_histogram.sql) with changes from internal team review. After a final review pass on this PR, the remaining queries will be updated with similar methods.

Summary of changes:
* bucket generation code for histograms [adjusted so powers of 10 don't fall on bucket edges](https://github.com/m-lab/stats-pipeline/blob/random-sampling/statistics/queries/continent_histogram.sql#L4)
* selected rows are fingerprinted and sorted in a struct ([downloads](https://github.com/m-lab/stats-pipeline/blob/random-sampling/statistics/queries/continent_histogram.sql#L35-L46 - [uploads](https://github.com/m-lab/stats-pipeline/blob/random-sampling/statistics/queries/continent_histogram.sql#L122-L132))
* a [random row for each IP address in the set is selected](https://github.com/m-lab/stats-pipeline/blob/random-sampling/statistics/queries/continent_histogram.sql#L57-L61) and [used to compute daily statistics](https://github.com/m-lab/stats-pipeline/blob/random-sampling/statistics/queries/continent_histogram.sql#L69-L72)

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/stats-pipeline/42)

### CODE REVIEW

**Review by mattmathis — CHANGES_REQUESTED** [review]
Sorry I didn't do a better job with the informal review.   I think this will come out much cleaner.

> *[Reviewable](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-:-MS-ovls7NRSgquvGjkP:b-a08ckh)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @critzo)

___
*[statistics/queries/continent_histogram.sql, line 46 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-kuNgDjrVHQCNQU2c:-MS-kuNgDjrVHQCNQU2d:becjj9e) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L46)):*
> ```TSQL
>       ROUND(MAX(mbps),3) AS download_MAX,
>       ROUND(APPROX_QUANTILES(MinRTT, 100) [SAFE_ORDINAL(50)],3) AS download_minRTT_MED
>     ) AS stats
> ```

It would be better to do all of the stats after the random selection.  I would leave a raw_tests counter.
___
*[statistics/queries/continent_histogram.sql, line 68 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-nBjbA4ym-JQT87iL:-MS-nBjbA4ym-JQT87iM:b86jdh1) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L68)):*
> ```TSQL
> dl_stats_per_day AS (
>   SELECT
>     date, continent_code,
> ```

Replace stats.download_ med by random1.mbps everywhere
Move the above stats down here.

Using random2 for one is a good idea.   dl_day_log_avg_random2 is probably best.

Add a counter for number of tests included per geo.
___
*[statistics/queries/continent_histogram.sql, line 131 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-nrsg0NNIJKdfi29n:-MS-nrsh8MLSyDjnnFZ1:b-34vmyd) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L131)):*
> ```TSQL
>       ROUND(AVG(mbps),3) AS upload_AVG,
>       ROUND(MAX(mbps),3) AS upload_MAX,
>       ROUND(APPROX_QUANTILES(MinRTT, 100) [SAFE_ORDINAL(50)],3) AS upload_minRTT_MED
> ```

See above
___
*[statistics/queries/continent_histogram.sql, line 184 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-m7Vv4q9ZbKrc9Wl0:-MS-m7Vv4q9ZbKrc9Wl1:b-ddu2ls) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L184)):*
> ```TSQL
> JOIN dl_stats_per_day USING (date, continent_code)
> JOIN ul_stats_per_day USING (date, continent_code)
> ORDER BY date, continent_code, bucket_min
> ```

Minor style point: if you make this:
Results AS ( ... )  
SELECT * FROM Results
ORDER BY date, continent_code, bucket_min
It is easier for others to test/adapt

**Comment by critzo** [comment]
ack.

**Comment by critzo** [comment]
___
*[statistics/queries/continent_histogram.sql, line 46 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-kuNgDjrVHQCNQU2c:-MS37mOA5NWoYB-SQFEV:b-i3p3x9) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L46)):*
<details><summary><i>Previously, mattmathis (Matt Mathis) wrote…</i></summary><blockquote>

It would be better to do all of the stats after the random selection.  I would leave a raw_tests counter.
</blockquote></details>

Ok, but I'm not sure that's what we want. The stats aren't based on random selection, they're based on all the tests in a given day/geo per IP. 

In `*_stats_per_day` we select the log average for each day using a random row from `members` for each IP on a given day/geo, and use that for the histogram, but the stats themselves are for the total set, and are just passed through.

**Comment by critzo** [comment]
___
*[statistics/queries/continent_histogram.sql, line 46 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-kuNgDjrVHQCNQU2c:-MS3Ah-F6Y0JVzgghoqh:b-2ump2n) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L46)):*
<details><summary><i>Previously, critzo wrote…</i></summary><blockquote>

Ok, but I'm not sure that's what we want. The stats aren't based on random selection, they're based on all the tests in a given day/geo per IP. 

In `*_stats_per_day` we select the log average for each day using a random row from `members` for each IP on a given day/geo, and use that for the histogram, but the stats themselves are for the total set, and are just passed through.



</blockquote></details>

Another reason the stats are here, IIUC, is that we need to have stats per IP/day/geo. Moving this to `_stats_per_day` would just mean we also have to do the IP grouping there, and I think having it here is more readable. But if you think it's not correct, I think I need a suggested change in a commit.

**Comment by critzo** [comment]
___
*[statistics/queries/continent_histogram.sql, line 184 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-m7Vv4q9ZbKrc9Wl0:-MS3Bm7fCP9ZaRqbhrQu:b-pqov82) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L184)):*
<details><summary><i>Previously, mattmathis (Matt Mathis) wrote…</i></summary><blockquote>

Minor style point: if you make this:
Results AS ( ... )  
SELECT * FROM Results
ORDER BY date, continent_code, bucket_min
It is easier for others to test/adapt
</blockquote></details>

makes sense

**Comment by critzo** [comment]
___
*[statistics/queries/continent_histogram.sql, line 68 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-nBjbA4ym-JQT87iL:-MS377F_EcfrVNOMDWJZ:b-l9u4kt) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L68)):*
<details><summary><i>Previously, mattmathis (Matt Mathis) wrote…</i></summary><blockquote>

Replace stats.download_ med by random1.mbps everywhere
Move the above stats down here.

Using random2 for one is a good idea.   dl_day_log_avg_random2 is probably best.

Add a counter for number of tests included per geo.
</blockquote></details>

I think the reason @gfr10598 used `stats.download_med` is that he wanted the histogram metric to be the log average of the Median. The change you're suggesting would change the metric to the log average of a randomly selected row (if we leave stats selection above)

**Comment by critzo** [comment]
___
*[statistics/queries/continent_histogram.sql, line 131 at r1](https://reviewable.io/reviews/m-lab/stats-pipeline/42#-MS-nrsg0NNIJKdfi29n:-MS3Br7w4iXjJBbkjhhd:b-bxck1a) ([raw file](https://github.com/m-lab/stats-pipeline/blob/f709c11948a93b62108e06410543d3e7dbc047d6/statistics/queries/continent_histogram.sql#L131)):*
<details><summary><i>Previously, mattmathis (Matt Mathis) wrote…</i></summary><blockquote>

See above
</blockquote></details>

ack. repeat for uoloads. will address once we discuss questions above.

**Comment by critzo** [comment]
@mattmathis Thanks for reviewing. I left a couple questions, and maybe we can find a time to discuss.

File: statistics/queries/continent_histogram.sql [hunk 1]
```diff
@@ -18,61 +19,60 @@ dl_per_location AS (
...
     continent_code,
     ip,
-    MIN(mbps) AS download_MIN,
-    APPROX_QUANTILES(mbps, 100) [SAFE_ORDINAL(25)] AS download_Q25,
-    APPROX_QUANTILES(mbps, 100) [SAFE_ORDINAL(50)] AS download_MED,
-    AVG(mbps) AS download_AVG,
-    APPROX_QUANTILES(mbps, 100) [SAFE_ORDINAL(75)] AS download_Q75,
-    MAX(mbps) AS download_MAX,
-    APPROX_QUANTILES(MinRTT, 100) [SAFE_ORDINAL(50)] AS download_minRTT_MED
+    # This organizes all tests, in an arbitrary but repeatable order.
```
Comment (critzo) [inline statistics/queries/continent_histogram.sql:35]: @mattmathis suggests that this subquery should select one random row per IP/day/geo, and stats should be computed in a later subquery.

File: statistics/queries/continent_histogram.sql [hunk 2]
```diff
@@ -82,10 +82,9 @@ dl_histogram AS (
     CASE WHEN bucket_left = 0.31622776601683794 THEN 0
     ELSE bucket_left END AS bucket_min,
     bucket_right AS bucket_max,
-    COUNTIF(download_MED < bucket_right AND download_MED >= bucket_left) AS dl_samples_bucket,
-    COUNT(*) AS dl_samples_day,
-    COUNTIF(download_MED < bucket_right AND download_MED >= bucket_left) / COUNT(*) AS dl_frac
-  FROM dl_stats_perip_perday CROSS JOIN buckets
+    COUNTIF(stats.download_MED < bucket_right AND stats.download_MED >= bucket_left) AS dl_samples_bucket,
+    ROUND(COUNTIF(stats.download_MED < bucket_right AND stats.download_MED >= bucket_left) / COUNT(*), 3) AS dl_frac_bucket
+  FROM dl_random_perip_perday CROSS JOIN buckets
```
Comment (critzo) [inline statistics/queries/continent_histogram.sql:87]: The histogram should use the random rows for each IP/day/geo (mbps, not mbps_MED)

File: statistics/queries/continent_histogram.sql [hunk 3]
```diff
@@ -18,61 +19,60 @@ dl_per_location AS (
...
-    COUNT(*) AS dl_total_samples,
-    date,
-    continent_code
-  FROM dl_per_location_cleaned
-  GROUP BY
-    date,
-    continent_code
+    date, continent_code,
+    ROUND(POW(10,AVG(Safe.LOG10(stats.download_MED))),3) AS dl_day_log_avg_median,
+    ROUND(POW(10,AVG(Safe.LOG10(random1.MinRtt))),3) AS dl_min_rtt_day_log_avg,
```
Comment (critzo) [inline statistics/queries/continent_histogram.sql:70]: Calculate log avg twice for these metrics, using random1 and random2

Comment (critzo) [reply statistics/queries/continent_histogram.sql]: Done in [d516ad7](https://github.com/m-lab/stats-pipeline/pull/42/commits/d516ad7bdad668b0b71f348e532838ab21b1c438)

File: statistics/queries/continent_histogram.sql (hunk 2 above)
Comment (critzo) [reply statistics/queries/continent_histogram.sql:84]: Done in [d516ad7](https://github.com/m-lab/stats-pipeline/pull/42/commits/d516ad7bdad668b0b71f348e532838ab21b1c438)

File: statistics/queries/continent_histogram.sql (hunk 1 above)
Comment (critzo) [reply statistics/queries/continent_histogram.sql]: Done in [2706553](https://github.com/m-lab/stats-pipeline/pull/42/commits/270655378d2897ab096b46d44e92ef7485f8848c)

**Comment by critzo** [comment]
@mattmathis Latest commits implement the changes we discussed. Could you PTAL?

**Pull request merged**